### PR TITLE
Improvements to SVG. Fixes #2992.

### LINKED ISSUE #2992: cannot run sketch with this particular SVG file
In short sketch could not be built (java mode, processing v 3.0a5) with following SVG file at all
file: https://openclipart.org/detail/184667/2-euro-coin-by-frankes-184667

Build messages: 
```
Exception in thread "Animation Thread" java.lang.NullPointerException at 
processing.core.PShapeSVG.parseStyleAttributes(PShapeSVG.java:1334) at 
processing.core.PShapeSVG$Gradient.(PShapeSVG.java:1419) at 
processing.core.PShapeSVG$LinearGradient.(PShapeSVG.java:1441) at 
processing.core.PShapeSVG.parseChild(PShapeSVG.java:380) at 
processing.core.PShapeSVG.parseChildren(PShapeSVG.java:306) at processing.core.PShapeSVG.(PShapeSVG.java:295) at processing.core.PShapeSVG.parseChild(PShapeSVG.java:338) at 
processing.core.PShapeSVG.parseChildren(PShapeSVG.java:306) at processing.core.PShapeSVG.(PShapeSVG.java:295) at processing.core.PShapeSVG.(PShapeSVG.java:181) at 
processing.core.PGraphicsJava2D.loadShape(PGraphicsJava2D.java:1695) at 
processing.core.PGraphicsJava2D.loadShape(PGraphicsJava2D.java:1684) at 
processing.core.PApplet.loadShape(PApplet.java:11734) at test.setup(test.java:42) at 
processing.core.PApplet.handleDraw(PApplet.java:2365) at 
processing.core.PGraphicsJava2D.requestDraw(PGraphicsJava2D.java:245) at 
processing.core.PApplet.run(PApplet.java:2260) at java.lang.Thread.run(Unknown Source)
```

The problem is described in more detail here:
http://forum.processing.org/two/discussion/comment/31398#Comment_31398

### PR DESCRIPTION
**Pull request by GKFX**
Proper support for percentages, partial support for viewbox, full
support for named colors, allow gradients to specify color-stops using
style or direct attributes on &lt;stop&gt; (fix #2992), nicer error message on loading an
HTML document as SVG (since I assume it can happen to novices
downloading SVGs from the web).

~~Also unrelated edit to PApplet which disabled a doc comment which was
coming up in the wrong place.~~

### CODE REVIEW

File: core/src/processing/core/PShapeSVG.java [hunk 1]
```diff
@@ -406,16 +439,13 @@ protected PShape parseChild(XML elem) {
     } else if (name.equals("pattern")) {
       PGraphics.showWarning("Patterns are not supported.");
 
-    } else if (name.equals("stop")) {
```
Comment (benfry) [inline core/src/processing/core/PShapeSVG.java]: Don't remove this, there's documentation here for each item. If we wanted it to be a single if entry, we'd have done it that way in the first place.

File: core/src/processing/core/PShapeSVG.java [hunk 2]
```diff
@@ -1320,17 +1333,214 @@ void setColor(String colorText, boolean isFill) {
...
+    }
+  }
+
+
+  /**
+   * From <a href="http://www.w3.org/TR/SVG/single-page.html#types-ColorKeywords">the spec</a>.
+   */
+  static protected HashMap<String, Integer> colorNames;
+
+  static {
```
Comment (benfry) [inline core/src/processing/core/PShapeSVG.java]: Is this worth 150 lines of code in this file? What application creates SVGs that uses these X Window System colors from the mid-80s?

File: core/src/processing/core/PApplet.java [hunk 3]
```diff
@@ -3867,7 +3867,7 @@ static public final float min(float[] list) {
   }
 
 
-  /**
```
Comment (benfry) [inline core/src/processing/core/PApplet.java]: Please don't mix unrelated pull requests.

File: core/src/processing/core/PShapeSVG.java [hunk 4]
```diff
@@ -287,7 +330,10 @@ protected PShapeSVG(PShapeSVG parent, XML properties, boolean parseKids) {
 
     String transformStr = properties.getString("transform");
     if (transformStr != null) {
-      matrix = parseTransform(transformStr);
+      if (matrix == null)
```
Comment (benfry) [inline core/src/processing/core/PShapeSVG.java]: Always use braces. See https://github.com/processing/processing/wiki/Style-Guidelines

File: core/src/processing/core/PShapeSVG.java [hunk 5]
```diff
@@ -324,10 +365,7 @@ protected PShape parseChild(XML elem) {
     String name = elem.getName();
     PShapeSVG shape = null;
 
-    if (name == null) {
```
Comment (benfry) [inline core/src/processing/core/PShapeSVG.java]: Why is this removed? If you don't know why it's here, don't touch it.

File: core/src/processing/core/PApplet.java (hunk 3 above)
Comment (GKFX) [inline core/src/processing/core/PApplet.java]: Sorry – I was aware of the policy; I just thought you'd rather not receive a one-byte pull request for this. I'm more than happy to submit one if you want it.

File: core/src/processing/core/PShapeSVG.java (hunk 5 above)
Comment (GKFX) [inline core/src/processing/core/PShapeSVG.java]: Sorry, it looks like I briefly forgot what a null check was! I'll try to avoid doing daft things like that in future.

**Comment by GKFX** [comment]
OK, I've made the requested changes. Thanks for checking over it so quickly. (I deleted the original commit because it had the wrong email address attached.)

**Comment by benfry** [comment]
Thanks for this, I really appreciate you taking the time and fixing it up. 

Much better solution on the colors as well.
